autoscale/src/main/java/com/sequenceiq/periscope/service/MandatoryScalingAdjustmentService.java

```diff
@@ -0,0 +1,169 @@
+import com.sequenceiq.periscope.utils.ClusterUtils;
+import com.sequenceiq.periscope.utils.StackResponseUtils;
+import com.sequenceiq.periscope.utils.TimeUtil;
+
+@Service
+public class MandatoryScalingAdjustmentService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MandatoryScalingAdjustmentService.class);
+
+    @Inject
+    private ScalingEventSender scalingEventSender;
+
+    @Inject
+    private ClusterService clusterService;
+
+    @Inject
+    private StackResponseUtils stackResponseUtils;
+
+    @Inject
+    private YarnMetricsClient yarnMetricsClient;
+
+    @Inject
+    private YarnResponseUtils yarnResponseUtils;
+
+    private Cluster cluster;
+
+    private LoadAlert loadAlert;
+
+    private LoadAlertConfiguration loadAlertConfiguration;
+
+    private String policyHostGroup;
+
+    public void performMandatoryAdjustment(Long clusterId, String pollingUserCrn, StackV4Response stackResponse,
+            MandatoryScalingAdjustmentParameters scalingAdjustmentParameters) {
+
+        cluster = clusterService.findById(clusterId);
+        loadAlert = cluster.getLoadAlerts().iterator().next();
+        loadAlertConfiguration = loadAlert.getLoadAlertConfiguration();
+        policyHostGroup = loadAlert.getScalingPolicy().getHostGroup();
+
+        Map<String, String> hostFqdnsToInstanceId = stackResponseUtils.getCloudInstanceIdsForHostGroup(stackResponse, policyHostGroup);
+        List<String> servicesHealthyHostInstanceIds = stackResponseUtils.getCloudInstanceIdsWithServicesHealthyForHostGroup(stackResponse,
+                policyHostGroup);
+        List<String> stoppedHostIds = stackResponseUtils.getStoppedCloudInstanceIdsInHostGroup(stackResponse, policyHostGroup);
+
+        int stoppedNodeCount = stoppedHostIds.size();
+        int existingClusterNodeCount = stackResponse.getNodeCount() - stoppedNodeCount;
+
+        publishScalingEventIfNeeded(hostFqdnsToInstanceId, servicesHealthyHostInstanceIds, existingClusterNodeCount, stoppedHostIds, pollingUserCrn,
+                stackResponse, scalingAdjustmentParameters);
+    }
+
+    private void publishScalingEventIfNeeded(Map<String, String> hostFqdnsToInstanceId, List<String> hostIdsWithServicesHealthy, int existingClusterNodeCount,
+            List<String> stoppedHostIds, String pollingUserCrn, StackV4Response stackV4Response,
+            MandatoryScalingAdjustmentParameters scalingAdjustmentParameters) {
+
+        if (upscalable(scalingAdjustmentParameters.getMandatoryStopStartUpscaleAdjustment())) {
+            Integer targetScaleUpCount = scalingAdjustmentParameters.getMandatoryStopStartUpscaleAdjustment()
+                    .stream().map(adj -> IntStream.of(adj, loadAlertConfiguration.getMaxScaleUpStepSize()).min().getAsInt()).findFirst().orElse(0);
+            if (stoppedHostIds.isEmpty()) {
+                scalingEventSender.sendScaleUpEvent(loadAlert, existingClusterNodeCount, hostFqdnsToInstanceId.size(), hostIdsWithServicesHealthy.size(),
```

Comment:
How would we log something to the activity console indicating an operation is a result of a mandatory operation vs a load request ?
Probably part of a general future enhancement jira.

Reply:
Filed: https://jira.cloudera.com/browse/CB-18606 and included the change discussed